### PR TITLE
Truncated Gaussian Heat and Smoke Release Formulation for WRF-Fire

### PR DESCRIPTION
**Pull request by kasrash**
This commit adds a new Truncated Gaussian fire and heat distribution scheme in the atmosphere for WRF-Fire. This scheme overcomes the shortcomings of the current scheme and improves simulation results accuracy.

TYPE: new feature

KEYWORDS: WRF-Fire, Heat Release, Smoke, Plume, Gaussian

SOURCE: Kasra Shamsaei (University of Nevada, Reno), Tim Juliano, Branko Kosovic (NCAR/RAL)

DESCRIPTION OF CHANGES:
Problem:
The current fire heat distribution in the WRF-Fire does not conserve the heat fluxes from the fire. In this new scheme, the fire heat is conserved while distributed in the atmosphere. Additionally, the current smoke tracer release scheme in WRF-Fire was over-simplified.

Solution:
We introduce a Truncated Gaussian fire heat and smoke distribution scheme in WRF-Fire to overcome the mentioned limitation of fire heat distribution and have a higher fidelity model to release fire smoke tracers. Several namelist options are added in order for the user to choose the desired release scheme and adjust the parameter.

LIST OF MODIFIED FILES:
Registry/registry.fire
phys/module_fr_fire_atm.F
phys/module_fr_fire_driver.F
phys/module_fr_fire_driver_wrf.F

TESTS CONDUCTED: 
1. Do mods fix problem? How can that be demonstrated, and was that test conducted?
The new scheme is explained and its effects are analyzed against real-world measurements in the following paper for historic fires simulations:

- Shamsaei, K.; Juliano, T.W.; Roberts, M.; Ebrahimian, H.; Lareau, N.P.; Rowell, E.; Kosovic, B. The Role of Fuel Characteristics and Heat Release Formulations in Coupled Fire-Atmosphere Simulation. Fire 2023, 6, 264. https://doi.org/10.3390/fire6070264

RELEASE NOTE: Add new Truncated Gaussian scheme to distribute fire heat and smoke in the atmosphere.

### CODE REVIEW

**Comment by weiwangncar** [comment]
The regression test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Review by brankokosovic — DISMISSED** [review]

**Comment by weiwangncar** [comment]
@pedro-jm @dudhia Would you like to review this PR?

**Comment by pedro-jm** [comment]
@weiwangncar 
OK, I will look

**Comment by weiwangncar** [comment]
> @weiwangncar OK, I will look

@pedro-jm Can you take a look at this soon?

**Comment by pedro-jm** [comment]
@weiwangncar I am aiming to take a look this week.

File: phys/module_fr_fire_atm.F [hunk 1]
```diff
@@ -33,10 +34,23 @@ subroutine add_fire_tracer_emissions(    &
...
+
 ! local
 integer::isz1,jsz1,isz2,jsz2,ir,jr
 integer::i,j,ibase,jbase,i_f,ioff,j_f,joff
 real::avgw,emis,conv
+integer :: i_st,i_en,j_st,j_en
+
+!local for TG
+integer :: k,k_st,k_en
+real, dimension(ims:ime,kms:kme,jms:jme) :: prop_smk
```
Comment (pedro-jm) [inline phys/module_fr_fire_atm.F:53]: If prop_smk is a local array, it is better to define it in the tiles dimensions

File: phys/module_fr_fire_atm.F [hunk 2]
```diff
@@ -124,6 +168,8 @@ SUBROUTINE fire_tendency( &
    REAL :: fact_g, fact_c
    REAL :: alfg_i, alfc_i
 
+   REAL, DIMENSION( ims:ime,kms:kme,jms:jme ) :: prop_heat !proportion of heat to be released fro TG dist.
```
Comment (pedro-jm) [inline phys/module_fr_fire_atm.F:171]: If prop_heat is a local array, it is more standard in WRF to define it using tile indeces

File: phys/module_fr_fire_atm.F [hunk 3]
```diff
@@ -46,18 +60,44 @@ subroutine add_fire_tracer_emissions(    &
...
-               if (num_tracer >0)then
+               if (num_tracer > 0)then
+                  if (fire_smk_scheme .eq. 0)then
                      emis=avgw*fire_tracer_smoke*burnt_area_dt(i_f,j_f)*fgip(i_f,j_f)*1000/(rho(i,kts,j)*dz8w(i,kts,j)) ! g_smoke/kg_air
                      tracer(i,kts,j,p_fire_smoke)=tracer(i,kts,j,p_fire_smoke)+emis
-                endif
+                  
+                  else if (fire_smk_scheme .eq. 1)then
+                     do k = k_st,k_en
+                        emis=prop_smk(i,k,j)*avgw*fire_tracer_smoke*burnt_area_dt(i_f,j_f)*fgip(i_f,j_f)*1000/(rho(i,k,j)*dz8w(i,k,j)) ! g_smoke/kg_air
```
Comment (pedro-jm) [inline phys/module_fr_fire_atm.F:94]: it looks possible that if fire_sfc_flx is 0 prop_smk is not initialized. ??

File: Registry/registry.fire [hunk 4]
```diff
@@ -221,6 +221,17 @@ rconfig   integer fire_sprd_mdl          namelist,fire         max_domains
...
 rconfig   real    fire_ext_grnd          namelist,fire         max_domains           50.     -      "fire_ext_grnd"    "extinction depth of sfc fire heat"   "m"
 rconfig   real    fire_ext_crwn          namelist,fire         max_domains           50.     -      "fire_ext_crwn"    "extinction depth of crown fire heat" "m"
+#
+# ------------------------------------------------------------------------------------------------------------------------
+# variable for Truncated Gaussian dist.
+#
+rconfig   integer fire_sfc_flx           namelist,fire         max_domains           0       -      "fire_sfc_flx"     "compute flux div according to 0=exponential decay, 1=Truncated Gaussian distribution" ""
+rconfig   real    fire_heat_peak         namelist,fire         max_domains           0.      -      "fire_heat_peak"   "ONLY fire_sfc_flx=1, the peak heat release height for the Truncated Gaussian scheme" "m AGL"
+rconfig   real    fire_tg_ub             namelist,fire         max_domains           1000.   -      "fire_tg_ub"       "The upper bpund of the Truncated Gaussian scheme; the default typically works well" "m AGL"
```
Comment (pedro-jm) [inline Registry/registry.fire:230]: is this option, fire_tg_ub, also "ONLY fire_sfc_flx = 1"?

**Review by pedro-jm — COMMENTED** [review]
This looks good to me. I left a few comments to consider.

**Comment by weiwangncar** [comment]
@kasrash Can you address these comments and make changes?

File: phys/module_fr_fire_atm.F (hunk 1 above)
Comment (weiwangncar) [reply phys/module_fr_fire_atm.F]: changed.

File: phys/module_fr_fire_atm.F (hunk 2 above)
Comment (weiwangncar) [reply phys/module_fr_fire_atm.F]: Changed.

**Comment by weiwangncar** [comment]
@pedro-jm I changed dimension for array 'prop*' to tile dimensioned.

File: phys/module_fr_fire_atm.F (hunk 3 above)
Comment (pedro-jm) [reply phys/module_fr_fire_atm.F:94]: I think it is fine as it is.

File: Registry/registry.fire (hunk 4 above)
Comment (pedro-jm) [reply Registry/registry.fire:230]: it is also used by fire_smk_scheme so looks good.

**Review by pedro-jm — APPROVED** [review]

**Review by weiwangncar — APPROVED** [review]

File: phys/module_fr_fire_atm.F [hunk 5]
```diff
@@ -46,18 +60,44 @@ subroutine add_fire_tracer_emissions(    &
...
+   call tg_dist(ims,ime, kms,kme, jms,jme, &
+                i_st,i_en, j_st,j_en, k_st,k_en, dz8w, &
+                fire_smk_peak,fire_tg_ub,fire_smk_ext,z_at_w,zs, &
+                prop_smk)
+end if
+
+do j=j_st,j_en
     jbase=jtfs+jr*(j-jts)
-    do i=max(ids+1,its),min(ite,ide-2)
+    do i=i_st,i_st
```
[the remark this answers is not in the document]
Comment (twjuliano) [reply phys/module_fr_fire_atm.F:81]: I was reviewing this PR and noticed that my comment somehow never went through. @kasrash could you please comment on this? Thanks

**Pull request merged**